lib/src/session/async_item.rs

```diff
@@ -93,4 +97,62 @@ impl AsyncItem {
             Self::PendingReq(req) => req.await_ready().await,
         }
     }
+
+    pub fn is_ready(&mut self) -> bool {
+        self.await_ready().now_or_never().is_some()
+    }
+}
+
+impl From<PeekableTask<Response<Body>>> for AsyncItem {
+    fn from(req: PeekableTask<Response<Body>>) -> Self {
+        Self::PendingReq(req)
+    }
+}
+
+#[derive(Debug)]
+pub enum PeekableTask<T> {
+    Waiting(oneshot::Receiver<Result<T, Error>>),
+    Complete(Result<T, Error>),
+}
+
+impl<T: Send + 'static> PeekableTask<T> {
+    pub async fn spawn(fut: impl Future<Output = Result<T, Error>> + 'static + Send) -> Self {
+        let (sender, receiver) = oneshot::channel();
+        tokio::task::spawn(async move { sender.send(fut.await) });
+        Self::Waiting(receiver)
+    }
+
+    pub fn complete(t: T) -> Self {
+        PeekableTask::Complete(Ok(t))
+    }
+
+    /// Block until a response is ready.
+    pub async fn await_ready(&mut self) {
+        if let PeekableTask::Waiting(rx) = self {
+            if let Ok(v) = rx.await {
+                *self = PeekableTask::Complete(v)
+            } else {
+                // todo, not the correct error type
+                *self =
+                    PeekableTask::Complete(Err(Error::GuestError(wiggle::GuestError::PtrOverflow)));
```

Comment:
Since this branch represents the other end of the channel closing, that means the task holding the write end has disappeared without writing to it. The ordinary explanations for this are a timeout causing that task to be dropped, or the task erroring out in a way that it fails to write an `Err` to its oneshot sender, which would be a bug.

Viceroy doesn't currently have timeouts on session execution, so I believe the only explanation for ending up in this branch would be a bug. I think it's probably safe to then return an `Error::Other` like this:

```suggestion
                // todo, not the correct error type
                *self =
                    PeekableTask::Complete(Err(anyhow!("peekable task sender unexpectedly dropped").into()));
```